diff --git a/src/initialize.cpp b/src/initialize.cpp
index 1bee11d..7de8d0e 100644
--- a/src/initialize.cpp
+++ b/src/initialize.cpp
@@ -15,19 +15,19 @@ void initialize() {
   double init = timer.millis().convert(millisecond);
   macro::print("Starting Init Routine");
 
-  Autonomous Auton;
-  macro::print("Auton Set!");
+  // Autonomous Auton;
+  // macro::print("Auton Set!");
 
   // Class Initialization
   Odom odom;
   Chassis chassis(odom.getL(), odom.getR(), odom.getThetaDeg(), odom.getX(), odom.getY());
-  Rack rack;
-  Arm arm;
+  // Rack rack;
+  // Arm arm;
   Display Disp;
 
   // Roller Init
-  io::RollerL.set_brake_mode(MOTOR_BRAKE_HOLD);
-	io::RollerR.set_brake_mode(MOTOR_BRAKE_HOLD);
+  // io::RollerL.set_brake_mode(MOTOR_BRAKE_HOLD);
+	// io::RollerR.set_brake_mode(MOTOR_BRAKE_HOLD);
 
   // Sensor Init
   odom.calibrateGyro();
@@ -48,8 +48,8 @@ void initialize() {
   pros::Task b_display(Disp.start, NULL, "Display Controller");
   b_display.set_priority(TASK_PRIORITY_MIN);
 
-  pros::Task b_auton(Auton.start, NULL, "Auton Controller");
-  b_display.set_priority(TASK_PRIORITY_MIN);
+  // pros::Task b_auton(Auton.start, NULL, "Auton Controller");
+  // b_display.set_priority(TASK_PRIORITY_MIN);
 
   std::cout << "Tasks Initialized!" << std::endl;
 
@@ -60,9 +60,9 @@ void initialize() {
       .addInfo("X", 'd', odom.getX())
       .addInfo("Y", 'd', odom.getY());
 
-  Disp.addInfo("Rack", 'i', rack.getPot())
-      .addInfo("Arm", 'd', arm.getPos())
-      .addInfo("Arm Limit", 'b', arm.getLimit());
+  // Disp.addInfo("Rack", 'i', rack.getPot())
+  //     .addInfo("Arm", 'd', arm.getPos())
+  //     .addInfo("Arm Limit", 'b', arm.getLimit());
 
   Disp.cleanUp();
 
diff --git a/src/kari/display/displayController.cpp b/src/kari/display/displayController.cpp
index 08fb2ce..d47b535 100644
--- a/src/kari/display/displayController.cpp
+++ b/src/kari/display/displayController.cpp
@@ -46,7 +46,7 @@ static lv_res_t btn_click_action(lv_obj_t * btn) {
   switch(id) {
     case 1: odom.zero(); break;
     case 2: odom.reset(); break;
-    case 3: arm.tarePos(); break;
+    // case 3: arm.tarePos(); break;
 
     default: break;
   }
